libraries/mempool/mempool.cpp

```diff
@@ -22,435 +20,535 @@ namespace koinos::mempool {
-            );
-
-            transaction_iterator = _pending_transactions.emplace( account_nonce_iterator->iterator, std::move( pending_transaction ) );
-         }
-         else
-         {
-            _pending_transactions.push_back( std::move( pending_transaction ) );
-            transaction_iterator = --_pending_transactions.end();
-         }
-
-         KOINOS_ASSERT( transaction_iterator != _pending_transactions.end(), pending_transaction_insertion_failure, "failed to insert transaction with id: ${id}", ("id", id) );
-
-         auto rval = _pending_transaction_idx.emplace( pending_transaction_iterator_wrapper{ transaction_iterator } );
-         if ( !rval.second )
-         {
-            _pending_transactions.erase( transaction_iterator );
-            KOINOS_ASSERT( false, pending_transaction_insertion_failure, "failed to insert transaction with id: ${id}", ("id", id) );
-         }
-      }
+   auto address_resource_bytes = util::converter::as< std::string >( arr );
+   node->put_object( space::address_resources(), transaction.header().payer(), &address_resource_bytes );
 
-      auto& account_idx = _account_resources_idx.get< by_account >();
-      auto it = account_idx.find( payer );
+   return rc_used;
+}
 
-      if ( it == account_idx.end() )
-      {
-         _account_resources_idx.insert( account_resources_object {
-            .account       = payer,
-            .resources     = max_payer_rc - rc_limit,
-            .max_resources = max_payer_rc,
-            .time          = time
-         } );
-      }
-      else
+uint64_t mempool_impl::add_pending_transaction(
+   const protocol::transaction& transaction,
+   std::chrono::system_clock::time_point time,
+   uint64_t max_payer_rc,
+   uint64_t disk_storaged_used,
+   uint64_t network_bandwidth_used,
+   uint64_t compute_bandwidth_used )
+{
+   uint64_t rc_used = 0;
+
+   auto lock = _db.get_shared_lock();
+   auto nodes = _db.get_fork_heads( lock );
+
+   try
+   {
+      std::vector< state_db::anonymous_state_node_ptr > anonymous_state_nodes;
+
+      for ( auto block_node : nodes )
       {
-         int128_t max_resource_delta = int128_t( max_payer_rc ) - int128_t( it->max_resources );
-         int128_t new_resources = int128_t( it->resources ) + max_resource_delta - int128_t( rc_limit );
+         auto pending_node = relevant_node( block_node->id(), lock );
+         assert( pending_node );
+         auto node = pending_node->create_anonymous_node();
+         anonymous_state_nodes.push_back( node );
 
-         account_idx.modify( it, [&]( account_resources_object& aro )
-         {
-            aro.max_resources = max_payer_rc;
-            aro.resources     = new_resources.convert_to< uint64_t >();
-            aro.time          = time;
-         } );
+         auto rc = add_pending_transaction_to_fork( node, transaction, time, max_payer_rc, disk_storaged_used, network_bandwidth_used, compute_bandwidth_used );
```

Comment:
An interesting implication.

In order for this to succeed, the pending transaction must not exceed max account rc on ALL forks. Is this desirable?